### ISSUE TITLE
Rollback Slave as Master Facility in Replication Manager

### ISSUE THREAD

**Issue by Manoharan-NMS** [signal18/replication-manager#345]
Configuration Details:
File Name: /etc/replication-manager/cluster.d/cluster1.toml

db-servers-hosts = "x.x.x.a:3306,x.x.x.b:3306,x.x.x.c:3306"
db-servers-prefered-master = "x.x.x.a:3306,x.x.x.b:3306"
db-servers-ignored-hosts="x.x.x.c:3306"

switchover-max-slave-delay = 30
failover-max-slave-delay = 30

**Observations**
**Default Condition:**

x.x.x.a:3306 - Master
x.x.x.b:3306 - Slave
x.x.x.c:3306 - Slave

In above configuration, Failover created in x.x.x.a:3306. Then x.x.x.b:3306 become Master.
After that Failover created in x.x.x.b:3306. Then x.x.x.c:3306 become Slave.

**Expected Result**
How to make x.x.x.c:3306 as Master with this condition?

x.x.x.a:3306 made as normal, after Failover. Expecting x.x.x.a:3306 as Master using Replication Manager.

**Actual Result**
Replication Manager keeps on restating. Not able to login.
**Logs:**
time="2021-05-01 16:35:25" level=info msg="No existing password encryption scheme" error="Key file does not exist"
time="2021-05-01 16:35:25" level=info msg="Failover in automatic mode" cluster=cluster1
time="2021-05-01 16:35:25" level=info msg="No existing password encryption scheme in LoadAPIUsers" cluster=cluster1
time="2021-05-01 16:35:26" level=info msg="Starting HTTPS & JWT API on 0.0.0.0:60005"

Related Issue: [https://github.com/signal18/replication-manager/issues/337]()

**Comment by svaroqui** [comment]
Thanks for reporting i was able to reproduce that issue and  working on it , indeed found some code  not adapted  for multiple db-servers-prefered-master. I'll keep you posted when i think the code is fixed.

**Comment by Manoharan-NMS** [comment]
> Thanks for reporting i was able to reproduce that issue and working on it , indeed found some code not adapted for multiple db-servers-prefered-master. I'll keep you posted when i think the code is fixed.

Thanks for your prompt response.

**Comment by svaroqui** [comment]
> > Thanks for reporting i was able to reproduce that issue and working on it , indeed found some code not adapted for multiple db-servers-prefered-master. I'll keep you posted when i think the code is fixed.
> 
> Thanks for your prompt response.

In the current code this should work for switchover 
db-servers-hosts = "x.x.x.a:3306,x.x.x.b:3306,x.x.x.c:3306"
db-servers-prefered-master = "x.x.x.a:3306"
db-servers-ignored-hosts="x.x.x.c:3306"

For failover  things are more complicated as it always pick the most up to date replica as the a new leader 
I have some code to switchover again if not in prefered leader list but it's not triggered and may have to get an additional option to enable it

**Comment by Manoharan-NMS** [comment]
> > > Thanks for reporting i was able to reproduce that issue and working on it , indeed found some code not adapted for multiple db-servers-prefered-master. I'll keep you posted when i think the code is fixed.
> > 
> > 
> > Thanks for your prompt response.
> 
> In the current code this should work for switchover
> db-servers-hosts = "x.x.x.a:3306,x.x.x.b:3306,x.x.x.c:3306"
> db-servers-prefered-master = "x.x.x.a:3306"
> db-servers-ignored-hosts="x.x.x.c:3306"
> 
> For failover things are more complicated as it always pick the most up to date replica as the a new leader
> I have some code to switchover again if not in prefered leader list but it's not triggered and may have to get an additional option to enable it

The above switchover option I verified. Its working fine. My concern is Failover only

**Comment by Manoharan-NMS** [comment]
I've MariaDB 10.4 with Master,Slave,Slave replication.

1) By default setup as below.
created new database in DB1.
All DB's are replicated in DB2,DB3

DB Status:

DB1: Master
DB2: Slave
DB3: Slave

All replications are happened properly.

2) Now performing failover in DB1.
created new database in DB2
All DB's are replicated in DB3


DB Status:

DB1: Failed
DB2: Master
DB3: Slave

3) Now performing failover in DB2.

DB Status:

DB1: Failed
DB2: Failed
DB3: Slave

Preparing the DB3(Slave) to be Standalone Mode

STOP ALL SLAVES;
RESET SLAVE ALL;
SHOW MASTER STATUS;
SELECT @@global.gtid_binlog_pos;
SET @@global.read_only=0;


DB Status:

DB1: Failed
DB2: Failed
DB3: Standalone

created new database in DB3

4) Changing the Old Master(DB1) to be a Slave

set @@global.read_only=1;
STOP ALL SLAVES;
RESET SLAVE;
CHANGE MASTER TO MASTER_HOST = '<DB3 HOST>',MASTER_PORT = <DB3 PORT>,MASTER_USER = 'db3user',MASTER_PASSWORD = 'db3password', master_use_gtid = current_pos, MASTER_LOG_FILE='<DB3 LOG FILE NAME>', MASTER_LOG_POS=<DB3 LOG FILE POSITION>;

START SLAVE;

Expected Result:
All DB's created in DB2,DB3 should get reflected.

Actual Result:
DB's created in DB3 is reflected. But DB's created in DB2 not reflected.

Is there any way to sync DB2 data.?

**Comment by svaroqui** [comment]
Yes this scenario should be covered without you doing any action on the databases , you may confuse replication-manager that should do the rejoin for you when the  old masters show up again , let me finish with fixing this issue first than we can plan a meeting to explain more in details the rejoin scenario that can be from flashback, backups , script

**Comment by nyxneuf** [comment]
My Test System Scenario.
CentOS7 / MySQL 5.7 / Master, Slave1, Slave2 
SRM :  replication-manager-osc-2.1.4-1.x86_64

db-servers-hosts = "192.168.50.11:3307,192.168.50.12:3307,192.168.50.13:3307"
db-servers-prefered-master = "192.168.50.11:3307"
db-servers-ignored-hosts = "192.168.50.13:3307"

replicaion-manaher swichover is success! No Problem. 

Master is Fail-over is Slave Rejoin failed!

**Comment by Manoharan-NMS** [comment]
> My Test System Scenario.
> CentOS7 / MySQL 5.7 / Master, Slave1, Slave2
> SRM : replication-manager-osc-2.1.4-1.x86_64
> 
> db-servers-hosts = "192.168.50.11:3307,192.168.50.12:3307,192.168.50.13:3307"
> db-servers-prefered-master = "192.168.50.11:3307"
> db-servers-ignored-hosts = "192.168.50.13:3307"
> 
> replicaion-manaher swichover is success! No Problem.
> 
> Master is Fail-over is Slave Rejoin failed!

When the fix can be available for Failover.

**Comment by tanji** [comment]
Fixed in 2.1.5 release

**Comment by svaroqui** [comment]
For the rejoin old master after failover  10.6 is promising using semisync  https://jira.mariadb.org/browse/MDEV-21117  long awaiting since that replication-manager need flashback

**Comment by Manoharan-NMS** [comment]
> Fixed in 2.1.5 release

Thanks @tanji  for your valuable feedback.
I'll verify in 2.1.5 version & update you

Link: http://ci.signal18.io/mrm/builds/tags/v2.1.5/replication-manager-osc-2.1.5-1.x86_64.rpm

**Comment by Manoharan-NMS** [comment]
> For the rejoin old master after failover 10.6 is promising using semisync https://jira.mariadb.org/browse/MDEV-21117 long awaiting since that replication-manager need flashback

Thanks for your information.

**Comment by Manoharan-NMS** [comment]
> Fixed in 2.1.5 release

I've MariaDB 10.4 with Master,Slave1,Slave2 replication.

Test Setup Details:
DB1 (Default Master): 192.168.9.42, Port: 3315
DB2 (Default Slave1): 192.168.9.42, Port: 3316
DB3 (Default Slave2): 192.168.9.42, Port: 3317

1) By default setup as below.

Conf File path: /etc/replication-manager/cluster.d/cluster1.toml

Enabled Conf File parameters

db-servers-hosts = "192.168.9.42:3315,192.168.9.42:3316,192.168.9.42:3317"
db-servers-prefered-master = "192.168.9.42:3315,192.168.9.42:3316"
db-servers-credential = "repli:repli"
db-servers-connect-timeout = 1
replication-credential = "repli:repli"
db-servers-ignored-hosts="192.168.9.42:3317"



created new database in DB1.
All DB's are replicated in DB2,DB3

Replication Manager Status:

DB1: Master
DB2: Slave
DB3: Slave

All replications are happened properly.

2) Now performing failover in DB1.

created new database in DB2
All DB's are replicated in DB3


Replication Manager Status:

DB1: Failed
DB2: Master
DB3: Slave

3) Now performing failover in DB2.

Replication Manager Status:

DB1: Failed
DB2: Failed
DB3: Slave

Now stopped Replication Manager,Enabled below configuration & restarted replication manager.

Conf File path: /etc/replication-manager/cluster.d/cluster1.toml

Enabled Conf File parameters

db-servers-hosts = "192.168.9.42:3315,192.168.9.42:3316,192.168.9.42:3317"
db-servers-prefered-master = "192.168.9.42:3317"
db-servers-credential = "repli:repli"
db-servers-connect-timeout = 1
replication-credential = "repli:repli"
db-servers-ignored-hosts="192.168.9.42:3316,192.168.9.42:3315"


Replication Manager Status:

DB1: Failed
DB2: Failed
DB3: Master

created new database in DB3

4) Changing the Last Updated Master(DB2) to be a Slave

Now started DB Instance at DB2

Replication Manager Status:

DB1: Failed
DB2: Slave
DB3: Master

All DB's created in DB3 is replicated in DB2.

5) Changing the Last known Master(DB2) to be a Master

Now stopped Replication Manager,Enabled below configuration & restarted replication manager.

Conf File path: /etc/replication-manager/cluster.d/cluster1.toml

Enabled Conf File parameters

db-servers-hosts = "192.168.9.42:3315,192.168.9.42:3316,192.168.9.42:3317"
db-servers-prefered-master = "192.168.9.42:3316"
db-servers-credential = "repli:repli"
db-servers-connect-timeout = 1
replication-credential = "repli:repli"
db-servers-ignored-hosts="192.168.9.42:3315"


Perform Switchover Now. Based on above configurations, DB2 become Master.

Replication Manager Status:

DB1: Failed
DB2: Master
DB3: Slave

6) Changing the Default Master(DB1) to be a Slave from Last known Master (DB2)

Now stopped Replication Manager,Enabled below configuration & restarted replication manager.

Conf File path: /etc/replication-manager/cluster.d/cluster1.toml

Enabled Conf File parameters


db-servers-hosts = "192.168.9.42:3315,192.168.9.42:3316,192.168.9.42:3317"
db-servers-prefered-master = "192.168.9.42:3315,192.168.9.42:3316"
db-servers-credential = "repli:repli"
db-servers-connect-timeout = 1
replication-credential = "repli:repli"
db-servers-ignored-hosts="192.168.9.42:3317"


DB1: Slave
DB2: Master
DB3: Slave

Expected Result:
In DB2, all DB's are in sync.(All DB's created in DB3 are available).
In DB1 also all DB's available in DB2 should be expected.

Actual Result:
DB1 become slave as expected. But DB's created in DB3 are not reflected in DB1.



**What's the way to sync missing data's ?**

@tanji All cases (From 1 to 5) are working fine. Thanks for your prompt response.

Please look into the above scenario(Case 6). During Failover data rollback problem is there.

**Comment by Manoharan-NMS** [comment]
> > Fixed in 2.1.5 release
> 
> I've MariaDB 10.4 with Master,Slave1,Slave2 replication.
> 
> Test Setup Details:
> DB1 (Default Master): 192.168.9.42, Port: 3315
> DB2 (Default Slave1): 192.168.9.42, Port: 3316
> DB3 (Default Slave2): 192.168.9.42, Port: 3317
> 
> 1. By default setup as below.
> 
> Conf File path: /etc/replication-manager/cluster.d/cluster1.toml
> 
> Enabled Conf File parameters
> 
> db-servers-hosts = "192.168.9.42:3315,192.168.9.42:3316,192.168.9.42:3317"
> db-servers-prefered-master = "192.168.9.42:3315,192.168.9.42:3316"
> db-servers-credential = "repli:repli"
> db-servers-connect-timeout = 1
> replication-credential = "repli:repli"
> db-servers-ignored-hosts="192.168.9.42:3317"
> 
> created new database in DB1.
> All DB's are replicated in DB2,DB3
> 
> Replication Manager Status:
> 
> DB1: Master
> DB2: Slave
> DB3: Slave
> 
> All replications are happened properly.
> 
> 1. Now performing failover in DB1.
> 
> created new database in DB2
> All DB's are replicated in DB3
> 
> Replication Manager Status:
> 
> DB1: Failed
> DB2: Master
> DB3: Slave
> 
> 1. Now performing failover in DB2.
> 
> Replication Manager Status:
> 
> DB1: Failed
> DB2: Failed
> DB3: Slave
> 
> Now stopped Replication Manager,Enabled below configuration & restarted replication manager.
> 
> Conf File path: /etc/replication-manager/cluster.d/cluster1.toml
> 
> Enabled Conf File parameters
> 
> db-servers-hosts = "192.168.9.42:3315,192.168.9.42:3316,192.168.9.42:3317"
> db-servers-prefered-master = "192.168.9.42:3317"
> db-servers-credential = "repli:repli"
> db-servers-connect-timeout = 1
> replication-credential = "repli:repli"
> db-servers-ignored-hosts="192.168.9.42:3316,192.168.9.42:3315"
> 
> Replication Manager Status:
> 
> DB1: Failed
> DB2: Failed
> DB3: Master
> 
> created new database in DB3
> 
> 1. Changing the Last Updated Master(DB2) to be a Slave
> 
> Now started DB Instance at DB2
> 
> Replication Manager Status:
> 
> DB1: Failed
> DB2: Slave
> DB3: Master
> 
> All DB's created in DB3 is replicated in DB2.
> 
> 1. Changing the Last known Master(DB2) to be a Master
> 
> Now stopped Replication Manager,Enabled below configuration & restarted replication manager.
> 
> Conf File path: /etc/replication-manager/cluster.d/cluster1.toml
> 
> Enabled Conf File parameters
> 
> db-servers-hosts = "192.168.9.42:3315,192.168.9.42:3316,192.168.9.42:3317"
> db-servers-prefered-master = "192.168.9.42:3316"
> db-servers-credential = "repli:repli"
> db-servers-connect-timeout = 1
> replication-credential = "repli:repli"
> db-servers-ignored-hosts="192.168.9.42:3315"
> 
> Perform Switchover Now. Based on above configurations, DB2 become Master.
> 
> Replication Manager Status:
> 
> DB1: Failed
> DB2: Master
> DB3: Slave
> 
> 1. Changing the Default Master(DB1) to be a Slave from Last known Master (DB2)
> 
> Now stopped Replication Manager,Enabled below configuration & restarted replication manager.
> 
> Conf File path: /etc/replication-manager/cluster.d/cluster1.toml
> 
> Enabled Conf File parameters
> 
> db-servers-hosts = "192.168.9.42:3315,192.168.9.42:3316,192.168.9.42:3317"
> db-servers-prefered-master = "192.168.9.42:3315,192.168.9.42:3316"
> db-servers-credential = "repli:repli"
> db-servers-connect-timeout = 1
> replication-credential = "repli:repli"
> db-servers-ignored-hosts="192.168.9.42:3317"
> 
> DB1: Slave
> DB2: Master
> DB3: Slave
> 
> Expected Result:
> In DB2, all DB's are in sync.(All DB's created in DB3 are available).
> In DB1 also all DB's available in DB2 should be expected.
> 
> Actual Result:
> DB1 become slave as expected. But DB's created in DB3 are not reflected in DB1.
> 
> **What's the way to sync missing data's ?**

@tanji   Please look into the above scenario during Failover rollback

**Comment by svaroqui** [comment]
Are you using gtid ?

**Comment by Manoharan-NMS** [comment]
> Are you using gtid ?

Yes. I'm using GTID only

**Comment by svaroqui** [comment]
Yes it looks like the gtid from the last crash is used to rejoin but in GTID it should not be forced indeed

**Comment by Manoharan-NMS** [comment]
> Yes it looks like the gtid from the last crash is used to rejoin but in GTID it should not be forced indeed

Default Condition:
DB1: Master
DB2: Slave
DB3: Slave

DB's created in DB1. Its reflected in DB2,DB3

Failover performed in DB1. DB2 become Master.
Now DB's created in DB2. Its reflected in DB3.

Failover performed in DB2. DB3 become master, after conf change in replication manager & replication manager restart.
Now DB's created in DB3. Not yet reflected. Because DB1 & DB2 are DOWN.

Now DB2 made UP & become Slave.
All DB's created in DB3 are available in DB2.

So DB2 - Slave has updated data.
Now performed switchover from DB3 to DB2, after conf change in replication manager & performed replication manager restart.

DB2 become master & it has all data's.

Now DB1 become UP & acted as Slave. Also pointed to DB2 master.

Actual Result:
The DB's created in DB3 is missing in DB1 become slave. But the same data available in DB2 & DB3.

How to resolve the problem. Any manual steps are there or any conf change required in replication manager.?

**Comment by svaroqui** [comment]
Can you pass me the log of this scenario , it's because rejoining a master but no the old master also log-level=3

**Comment by svaroqui** [comment]
For us to fixe this issue we need to understand what is the difference between a fresh server that is coming in for reseed or such scenario , i think you can  enable reseed from mysqldump 10.4 to see if it fixes it please produce a backup per node, also provide a correct mysqldump path parameter not the default one used by replication-manager

**Comment by Manoharan-NMS** [comment]
> For us to fixe this issue we need to understand what is the difference between a fresh server that is coming in for reseed or such scenario , i think you can enable reseed from mysqldump 10.4 to see if it fixes it please produce a backup per node, also provide a correct mysqldump path parameter not the default one used by replication-manager

Okay. I'll test & update you in some time. 

**Is there any possibilities to fix the issue in future release ? Because Master-Slave1-Slave2 this is one of the critical scenario**

**Comment by Manoharan-NMS** [comment]
> Can you pass me the log of this scenario , it's because rejoining a master but no the old master also log-level=3

I'll enable logger level 3 & provide you soon.

**Comment by svaroqui** [comment]
> > For us to fixe this issue we need to understand what is the difference between a fresh server that is coming in for reseed or such scenario , i think you can enable reseed from mysqldump 10.4 to see if it fixes it please produce a backup per node, also provide a correct mysqldump path parameter not the default one used by replication-manager
> 
> Okay. I'll test & update you in some time.
> 
> **Is there any possibilities to fix the issue in future release ? Because Master-Slave1-Slave2 this is one of the critical scenario**

Sure we will fix it at some point we have >100 clusters in production with master-slave-slave but the scenario you found is not high priority, and while in the past that scenario was covered introduction of reseed feature break it i think , support or direct contribution PR can help  to speed up the fix :)

**Comment by Manoharan-NMS** [comment]
> Can you pass me the log of this scenario , it's because rejoining a master but no the old master also log-level=3

Logs attached for given scenario.
[replication_logs_04062021.zip](https://github.com/signal18/replication-manager/files/6597745/replication_logs_04062021.zip)

**Comment by svaroqui** [comment]
Here is my 3 servers directory 
drwxr-xr-x   9 apple  staff   306B  4 jui 13:13 127.0.0.1_3317/
drwxr-xr-x   9 apple  staff   306B  4 jui 13:00 127.0.0.1_3318/
drwxr-xr-x   9 apple  staff   306B  3 jui 11:23 127.0.0.1_3319/
failed 127.0.0.1_3317  && 127.0.0.1_3318
restart 127.0.0.1_3318
restart repman 
restart  127.0.0.1_3317

I'm not able to reproduce 

```
-rw-r--r--   1 apple  staff   342B  3 jui 12:17 failover.20210603121713.json
-rw-r--r--   1 apple  staff   342B  4 jui 12:53 failover.20210604125330.json
-rw-r--r--   1 apple  staff   342B  4 jui 12:54 failover.20210604125427.json
-
```

So that's because you add an extra failover to come back to 127.0.0.1_3318 (b) and so erasing the previous failover state tacking.  So b indeed get 2 failover before the old master a come back taking the last failover info  of b to rejoin a and of course miss some events  

That scenario is of course possible  but the chance it happens in real life is very low

**Comment by Manoharan-NMS** [comment]
> Here is my 3 servers directory
> drwxr-xr-x 9 apple staff 306B 4 jui 13:13 127.0.0.1_3317/
> drwxr-xr-x 9 apple staff 306B 4 jui 13:00 127.0.0.1_3318/
> drwxr-xr-x 9 apple staff 306B 3 jui 11:23 127.0.0.1_3319/
> failed 127.0.0.1_3317 && 127.0.0.1_3318
> restart 127.0.0.1_3318
> restart repman
> restart 127.0.0.1_3317
> 
> I'm not able to reproduce
> 
> ```
> -rw-r--r--   1 apple  staff   342B  3 jui 12:17 failover.20210603121713.json
> -rw-r--r--   1 apple  staff   342B  4 jui 12:53 failover.20210604125330.json
> -rw-r--r--   1 apple  staff   342B  4 jui 12:54 failover.20210604125427.json
> -
> ```
> 
> So that's because you add an extra failover to come back to 127.0.0.1_3318 (b) and so erasing the previous failover state tacking. So b indeed get 2 failover before the old master a come back taking the last failover info of b to rejoin a and of course miss some events
> 
> That scenario is of course possible but the chance it happens in real life is very low

Thanks for your prompt response. Once you fixed the issue, please notify.

**Comment by Manoharan-NMS** [comment]
Any new fixes available in v2.1.6 tag @svaroqui 

Reference URL: http://ci.signal18.io/mrm/builds/tags/v2.1.6/
